### ISSUE TITLE
[Parity 8/16] Vision/audio kernels - grid_sample, interpolate, NMS, roi_align, STFT, MelSpectrogram, MFCC

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet.Tensors#217]
Part of epic #209.

## Summary

Vision and audio workloads rely on a specific kernel surface — `grid_sample`, `affine_grid`, `interpolate`, `pad`, NMS, roi_align/pool, box_iou, STFT, MelSpectrogram, MFCC. Most of these live in `torch.nn.functional` or `torchvision` / `torchaudio`. We're missing nearly all of them.

This issue owns the kernels; integration with model layers happens upstream in AiDotNet core.

### Geometry / sampling
- `F.grid_sample(input, grid, mode, padding_mode, align_corners)` — bilinear, nearest, bicubic + reflect/zeros/border padding
- `F.affine_grid(theta, size, align_corners)`
- `F.interpolate(input, size, scale_factor, mode, align_corners, antialias)` — nearest, linear, bilinear, bicubic, trilinear, area
- `F.pad(input, pad, mode)` — constant, reflect, replicate, circular
- `F.upsample` (legacy)

### Vision (torchvision)
- `nms(boxes, scores, iou_threshold)`, `batched_nms`
- `box_iou`, `generalized_box_iou`, `complete_box_iou`, `distance_box_iou`
- `roi_align(input, boxes, output_size, spatial_scale, sampling_ratio, aligned)`
- `roi_pool(input, boxes, output_size, spatial_scale)`
- `ps_roi_align`, `ps_roi_pool`
- `deform_conv2d`
- `masks_to_boxes`, `box_convert`

### Audio (torchaudio)
- `Spectrogram`, `MelSpectrogram`, `MFCC`
- `AmplitudeToDB`, `MuLawEncoding/Decoding`, `ComputeDeltas`, `Vol`
- `Resample` (polyphase / sinc), `Fade`
- `TimeStretch`, `PitchShift` (phase vocoder on top of STFT)
- Filter banks: Mel, linear, triangular

## Current state in the repo

- `src/AiDotNet.Tensors/Images/` exists — directory listing not yet inspected but likely partial image I/O. Confirm before starting.
- Nothing specific to grid_sample, roi_align, NMS, Mel, MFCC.

## How we beat PyTorch

1. **AVX-512 `grid_sample` / `interpolate`** with channels-last (NHWC) + NCHWc awareness — PyTorch's CPU interp collapses to generic strided loops. Our kernels stay vectorized on the packed layout. Target 2× on bilinear upsample of NCHW tensors.
2. **Fused resize + normalize + permute** pass for ImageNet-style preprocessing — one kernel instead of three. PyTorch does three dispatch hops.
3. **Batched NMS that's branch-free**: SIMD-friendly sort + parallel suppression. PyTorch CPU NMS is single-threaded. Target 5× on batched detection workloads.
4. **`roi_align` on channels-last / NCHWc** — we stay in the packed layout end-to-end; PyTorch's `roi_align` only knows NCHW.
5. **MelSpectrogram / MFCC as a compiled plan**: STFT + mel-filter-bank mul + log gets fused into one `CompiledInferencePlan`. PyTorch runs three ops.
6. **GPU parity across all 7 backends** — PyTorch only has CUDA + MPS for these ops.
7. **Sub-byte input support**: `interpolate`/`pad` with int1/2/3/4/NF4/FP4 input tensors for quantized vision pipelines (unique).
8. **Deformable convolutions with our autodiff** (including higher-order via torch.func issue).

### Geometry / sampling (depends on FFT issue only for STFT)
- [ ] `GridSample` (bilinear/nearest/bicubic, reflect/zeros/border padding, align_corners)
- [ ] `AffineGrid`
- [ ] `Interpolate` (nearest/linear/bilinear/bicubic/trilinear/area + antialias)
- [ ] `Pad` (constant/reflect/replicate/circular)

### Vision
- [ ] `Nms`, `BatchedNms`
- [ ] `BoxIou`, `GeneralizedBoxIou`, `CompleteBoxIou`, `DistanceBoxIou`
- [ ] `RoIAlign`, `RoIPool`, `PsRoIAlign`, `PsRoIPool`
- [ ] `DeformConv2d`
- [ ] `MasksToBoxes`, `BoxConvert`
- [ ] `ImageDecode` / `ImageEncode` for JPEG/PNG/WebP (via native codec bindings — leverage what's already in `Images/`)

### Audio (depends on FFT issue)
- [ ] `Spectrogram`
- [ ] `MelSpectrogram`
- [ ] `MFCC`
- [ ] `AmplitudeToDB`, `MuLawEncoding`, `MuLawDecoding`, `ComputeDeltas`
- [ ] `Resample` (polyphase + sinc)
- [ ] `TimeStretch`, `PitchShift` (phase vocoder on STFT)
- [ ] Filter banks: Mel, linear, triangular

### Bonus
- [ ] Fused `ResizeNormalizePermute` ImageNet preprocessing kernel
- [ ] Sub-byte interpolate/pad
- [ ] NCHWc-aware `RoIAlign`

## Acceptance criteria

- Numerical parity vs PyTorch/torchvision/torchaudio reference per op.
- Gradcheck passes for differentiable ops (`grid_sample`, `interpolate`, `roi_align`, `affine_grid`).
- CPU + CUDA backends; others best-effort.
- `MelSpectrogram` equivalence with torchaudio within documented tolerance.
- Benchmark wins: `grid_sample` ≥ 2× on CPU, `BatchedNms` ≥ 5× CPU vs torchvision, `MelSpectrogram` as a compiled plan ≥ 1.5× torchaudio.

## Implementation notes

- Module under `src/AiDotNet.Tensors/Engines/Simd/Vision/` and `Engines/Simd/Audio/`.
- GPU kernels under `Engines/DirectGpu/*/Vision/` and `Audio/`.
- STFT / MelSpectrogram reuse FFT + Windows from the FFT issue.
- ImageDecode/Encode build on existing `Images/` codec bindings.
- Autograd registrations in `DifferentiableOps.cs`.

**Comment by ooples** [comment]
# Implementation plan — detailed

## 1. Current state

`src/AiDotNet.Tensors/Images/` directory exists with image I/O (unaudited depth). No grid_sample, roi_align, NMS, Mel, MFCC. STFT/Mel/MFCC call into #212 FFT module.

## 2. Scope — full inventory

### Geometry / sampling
```
GridSample(input, grid, mode="bilinear"|"nearest"|"bicubic",
           padding_mode="zeros"|"border"|"reflection", align_corners=false)
AffineGrid(theta, size, align_corners=false)
Interpolate(input, size=null, scale_factor=null, mode="nearest"|"linear"|"bilinear"|
            "bicubic"|"trilinear"|"area", align_corners=null, antialias=false)
Pad(input, pad, mode="constant"|"reflect"|"replicate"|"circular", value=0)
```

### Vision
```
Nms(boxes, scores, iou_threshold)
BatchedNms(boxes, scores, idxs, iou_threshold)
BoxIou / GeneralizedBoxIou / CompleteBoxIou / DistanceBoxIou(boxes1, boxes2)
RoIAlign(input, boxes, output_size, spatial_scale, sampling_ratio, aligned)
RoIPool(input, boxes, output_size, spatial_scale)
PsRoIAlign / PsRoIPool
DeformConv2d(input, offset, weight, bias, stride, padding, dilation, groups, mask?)
MasksToBoxes(masks)
BoxConvert(boxes, in_fmt, out_fmt)  # xyxy ↔ xywh ↔ cxcywh
ImageDecode / ImageEncode (JPEG/PNG/WebP via native codec bindings)
```

### Audio (depends on #212)
```
Spectrogram(input, n_fft, hop_length, win_length, window, power, normalized, center,
            pad_mode, onesided)
MelSpectrogram(input, sample_rate, n_fft, hop, win, n_mels, f_min, f_max, window,
               power, normalized, center, pad_mode)
MFCC(input, sample_rate, n_mfcc, n_mels, dct_type, norm, log_mels, ...)
AmplitudeToDB(input, stype, top_db)
MuLawEncoding / MuLawDecoding(input, quantization_channels)
ComputeDeltas(input, win_length, mode)
Vol(input, gain, gain_type)
Resample(input, orig_freq, new_freq, lowpass_filter_width, rolloff, resampling_method,
         beta)  # polyphase + sinc
Fade(input, fade_in_len, fade_out_len, fade_shape)
TimeStretch(input, rate, n_freq)  # phase vocoder
PitchShift(input, sample_rate, n_steps, bins_per_octave, n_fft, win_length, hop_length)
Windows.Triangular / Mel / Linear filter banks
```

### Bonus
```
FusedResizeNormalizePermute  # ImageNet preprocessing one-kernel
Sub-byte Interpolate/Pad (int1/2/3/4/NF4/FP4 for quantized vision)
NCHWc-aware RoIAlign
```

**Total: ~40 ops.**

## 3. Win vectors

1. **AVX-512 `grid_sample` / `interpolate`** with NHWC + NCHWc awareness — PyTorch CPU interp collapses to generic strided loops. Target 2× bilinear upsample NCHW.
2. **Fused resize + normalize + permute**: one kernel for ImageNet preprocessing.
3. **Batched NMS branch-free**: SIMD-friendly sort + parallel suppression. Target 5× batched detection.
4. **`roi_align` on channels-last/NCHWc** — PyTorch only knows NCHW.
5. **MelSpectrogram/MFCC as a compiled plan**: STFT + mel-filterbank + log fused into one `CompiledInferencePlan`.
6. **GPU parity across all 7 backends** — PyTorch = CUDA + MPS only.
7. **Sub-byte input support** for `interpolate`/`pad` — unique.
8. **Deformable conv with our autodiff** incl. higher-order via #214.

## 4. Per-axis beat targets

| Axis | Target |
|---|---|
| CPU `grid_sample` bilinear NCHW | ≥2× PyTorch |
| CPU `interpolate` bilinear upsample | ≥2× PyTorch |
| Batched NMS CPU | ≥5× torchvision CPU |
| `MelSpectrogram` compiled plan | ≥1.5× torchaudio |
| `RoIAlign` channels-last | feature unique vs PyTorch |
| Sub-byte interpolate | feature unique |
| `DeformConv2d` + gradcheck | parity torchvision |
| All 6 GPU backends | CPU + CUDA required; others best-effort |

## 5. File inventory

**New (~90):**
- `src/AiDotNet.Tensors/Engines/Simd/Vision/` — GridSample, AffineGrid, Interpolate, Pad, Nms, BoxIou*, RoIAlign, RoIPool, DeformConv2d, MasksToBoxes, BoxConvert, FusedResizeNormalizePermute (~25)
- `src/AiDotNet.Tensors/Engines/Simd/Audio/` — Spectrogram, MelSpectrogram, MFCC, AmplitudeToDB, MuLaw*, ComputeDeltas, Vol, Resample, Fade, TimeStretch, PitchShift, FilterBanks (~15)
- GPU kernels per backend (~40)
- Sub-byte overloads (~10)

**Modified (~10):** `CpuEngine.cs`, `IEngine.cs`, GPU engines, `OpRegistry.cs`, `DifferentiableOps.cs`, `BackwardFunctions.cs`, `Images/` (reuse codec bindings).

**Tests (~35):** per-op parity vs PyTorch/torchvision/torchaudio, gradcheck for differentiables, device parity, sub-byte correctness, MelSpectrogram equivalence within documented tolerance.

**Benchmarks (~8):** `GridSampleBenchmarks`, `InterpolateBenchmarks`, `NmsBenchmarks`, `RoIAlignBenchmarks`, `MelSpectrogramBenchmarks`, `DeformConvBenchmarks`, `FusedPreprocessingBenchmarks`, `ResampleBenchmarks`.

## 6. Acceptance gate

1. Numerical parity vs PyTorch/torchvision/torchaudio per op.
2. Gradcheck for differentiable ops.
3. CPU + CUDA required; others best-effort.
4. MelSpectrogram equivalence within documented tolerance.
5. `grid_sample` ≥2× CPU; `BatchedNms` ≥5× CPU; `MelSpectrogram` compiled plan ≥1.5× torchaudio.
6. Sub-byte interpolate/pad works.

## 7. Timeline — ~6 weeks

W1 GridSample + AffineGrid + Interpolate + Pad (CPU SIMD).
W2 NMS family + BoxIou variants + RoIAlign/Pool.
W3 DeformConv2d + MasksToBoxes + BoxConvert + ImageDecode/Encode.
W4 Spectrogram + MelSpectrogram + MFCC + AmplitudeToDB + MuLaw + ComputeDeltas.
W5 Resample + Fade + TimeStretch + PitchShift + filter banks + fused kernels.
W6 GPU kernels across 6 backends + benchmarks + PR.

## 8. Dependencies

- **Blocks**: model-zoo demos (YOLO, DETR, Whisper, Wav2Vec).
- **Blocked by**: #212 (FFT — STFT/MelSpectrogram/MFCC), #210 (Unfold for im2col variants), #211 (affine_grid uses matrix_exp for stable forward path).

## 9. Risks

- **align_corners semantic subtleties** — PyTorch's behavior changed between versions; pin to 1.12+ convention.
- **DeformConv2d backward** CPU-slow in PyTorch; reference impl must be correct before optimizing.
- **PitchShift via phase vocoder** — artifacts at extreme shifts; match torchaudio tolerance.

---

*Part of epic #209. Plan ratified 2026-04-19.*
